core/src/main/java/com/linecorp/armeria/common/logging/DefaultLogWriter.java

```diff
@@ -0,0 +1,151 @@
+import static com.linecorp.armeria.common.logging.LogWriterBuilder.DEFAULT_RESPONSE_LOG_LEVEL_MAPPER;
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Predicate;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.util.SafeCloseable;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.TransientServiceOption;
+
+final class DefaultLogWriter implements LogWriter {
+
+    static final Logger defaultLogger = LoggerFactory.getLogger(DefaultLogWriter.class);
+
+    static final DefaultLogWriter DEFAULT =
+            new DefaultLogWriter(defaultLogger, DEFAULT_REQUEST_LOG_LEVEL_MAPPER,
+                                 DEFAULT_RESPONSE_LOG_LEVEL_MAPPER,
+                                 throwable -> false, LogFormatter.ofText());
+
+    private static boolean warnedNullRequestLogLevelMapper;
+    private static boolean warnedNullResponseLogLevelMapper;
+
+    private final Logger logger;
+    private final RequestLogLevelMapper requestLogLevelMapper;
+    private final ResponseLogLevelMapper responseLogLevelMapper;
+    private final Predicate<Throwable> responseCauseFilter;
+    private final LogFormatter logFormatter;
+
+    DefaultLogWriter(Logger logger, RequestLogLevelMapper requestLogLevelMapper,
+                     ResponseLogLevelMapper responseLogLevelMapper,
+                     Predicate<Throwable> responseCauseFilter, LogFormatter logFormatter) {
+        this.logger = logger;
+        this.requestLogLevelMapper = requestLogLevelMapper;
+        this.responseLogLevelMapper = responseLogLevelMapper;
+        this.responseCauseFilter = responseCauseFilter;
+        this.logFormatter = logFormatter;
+    }
+
+    @Override
+    public void logRequest(RequestOnlyLog log) {
+        requireNonNull(log, "log");
+        LogLevel requestLogLevel = requestLogLevelMapper.apply(log);
+        if (requestLogLevel == null) {
+            if (!warnedNullRequestLogLevelMapper) {
+                warnedNullRequestLogLevelMapper = true;
+                logger.warn("requestLogLevelMapper.apply() returned null; using default log level");
+            }
+            requestLogLevel = DEFAULT_REQUEST_LOG_LEVEL;
+        }
+        if (requestLogLevel.isEnabled(logger)) {
+            final RequestContext ctx = log.context();
+            if (log.requestCause() == null && isTransientService(ctx)) {
+                return;
+            }
+            try (SafeCloseable ignored = ctx.push()) {
+                // We don't log requestCause when it's not null because responseCause is the same exception when
+                // the requestCause is not null. That's way we don't have requestCauseSanitizer.
+                requestLogLevel.log(logger, logFormatter.formatRequest(log));
```

Comment:
> For example, a user might want to send a log to remote collector without using a logger API.

If we want to achieve the above, I guess we actually want to abstract this part.
e.g.
```
if (logger.isEnabled(requestLogLevel) {
  logAppender.append(logFormatter.formatRequest(log))
}
```

Having said this, I think this can be abstracted separately

Reply:
I was rather thinking something like
```java
@Override
public void logRequest(RequestOnlyLog log) {
  collectorWebClient.send(serializeLog(log));
}
```
that even a log appender isn't involved at all.